UI/MultiplayerLobbyWnd.cpp

```diff
@@ -159,19 +159,40 @@ namespace {
                     Select(0);
                 } else {
                     // For human players on host, have "Player", "Observer", and "Drop" options.  TODO: have "Ban" option.
-                    Insert(GG::Wnd::Create<TypeRow>(w, type_row_height, Networking::CLIENT_TYPE_HUMAN_PLAYER));   // "Human" display / option
-                    Insert(GG::Wnd::Create<TypeRow>(w, type_row_height, Networking::CLIENT_TYPE_HUMAN_OBSERVER)); // "Observer" display / option
-                    Insert(GG::Wnd::Create<TypeRow>(w, type_row_height, Networking::CLIENT_TYPE_HUMAN_MODERATOR));// "Moderator" display / option
+                    int row_player_type = -1;
+                    int row_observer_type = -1;
+                    int row_moderator_type = -1;
+                    int row_number = 0;
+                    if (client_type == Networking::CLIENT_TYPE_HUMAN_PLAYER ||
```

Comment:
What's the logic with the `||` here? Incomplete transition, or will there be player clients that don't have the player role...?

Reply:
Yes. It's a flexible system for any type of authorization configurations. Something like service account only to moderate games or where only authenticated players could play game and other could be only observers...

Reply:
Doesn't really address the question... why do you check both client type and roles set for it? When/why would an observer connection have the player role, and why wouldn't an observer connection have the observer role, for example? Presumably you want to get rid of the observer and moderator connection types and just have human connections that have various roles assigned...?

Reply:
I check both because it possible when host sets client type to player which doesn't have a role for it.
I don't plan to get rid of client types. It is used for current type of connection while roles determine possible set of types.

Reply:
I don't understand. How would a player not have a role "for" a client type...?

Roles determine possible set(s) of types...? I thought the roles determined what a connection could do in game / lobby?

Reply:
It could be if player get his client type from host. He could not do so itself because he doesn't have a role allows it but host have a role allows set client type for everyone.
Roles are what is allowed to do so roles for client types determines which client types player allowed to set.

Reply:
Then perhaps rename some roles to can-set-self-as-observer or similar.

Reply:
I could also add check before starting game and prevent it if some players doesn't have a roles to be allowed to use their client type.